### PR TITLE
Generator - support final fields

### LINKED ISSUE #193: Final fields can't be stored
**Description**
The scenario is adapting an existing model class to be compatible with ObjectBox. Assume the following `ToDo.dart` file which represents the model of `ToDo` item. Existing model already uses the reserved keyword `id` as a `String id` for another form of backend store such as Hive or InMemory list. Now to adapt this model to be compatible with ObjectBox terminology, I created a new field called `oboxId` with the `@Id` annottation.

**ToDo.dart**
```dart
import 'package:flutter_todo/commons/backend/base_store.dart';
import 'package:meta/meta.dart';
import 'package:objectbox/objectbox.dart';
import 'package:uuid/uuid.dart';

@Entity()

/// Model class to hold the ToDo Object.
class ToDo {
  /// Unique id of this todo to store in backend.
  /// Already used in Hive and InMemory lists.
  final String id;

  /// Id specifically required for ObjectBox, which must be of type int.
  @Id()
  int oboxId;

  /// Description of this todo.
  String description;

  /// A flag to denote if the to activity is complete
  bool completed;

  /// Due Date for this todo.
  DateTime dueDate;

  /// id of the user who created this todo.
  String userId;

  ToDo({
    @required this.description,
    this.dueDate,
    this.completed = false,
    @required this.userId,
  }) : id = Uuid().v4(); // Initializes id with a unique value.

  @override
  String toString() =>
      'Todo of user $userId with description $description and completed $completed';

  @override
  bool operator ==(o) =>
      o is ToDo &&
      o.description == description &&
      o.completed == completed &&
      o.dueDate == dueDate &&
      o.id == id &&
      o.oboxId == oboxId;

  @override
  int get hashCode =>
      description.hashCode ^
      userId.hashCode ^
      completed.hashCode ^
      dueDate.hashCode ^
      oboxId.hashCode;
}

```

**Basic info (please complete the following information):**
 - ObjectBox Dart version: [e.g. 0.7.0, latest version?]
 - Flutter/Dart SDK: [e.g. 1.20.0, latest version?]
 - Reproducibility: [e.g. occurred once only | occasionally without visible pattern | always]
 - Device: [e.g. Galaxy S20]
 - OS: [e.g. Android 10]

**To Reproduce**
1. Use the `ToDo.dart` as model class.
2. Run `flutter pub run build_runner build`
3. Then try to write a query in following format.

```dart
@override
  Future<ToDo> get(String id) async {
    final query = objBox.query(ToDo_.id.equals(id)).build(); // compiler throws error stating id is not  present in `ToDo_` type.
    
  }
```

**Expected behavior**
It is expected that the `id` field is still available for querying as it is required to fulfill already existing contracts in the code base.

**Additional Information**
<details>
<summary>pubspec.yaml</summary>

```yml
name: flutter_command_todo
description: A new Flutter project.

publish_to: "none"

version: 1.0.0+1

environment:
  sdk: ">=2.7.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  cupertino_icons: ^1.0.2
  flutter_command: ^0.9.3
  get_it: ^5.0.6
  uuid: ^2.2.2
  objectbox: ^0.11.0
  objectbox_flutter_libs: any

dev_dependencies:
  flutter_test:
    sdk: flutter
  build_runner:
  objectbox_generator: any

flutter:
  uses-material-design: true
```
</details>

------

**Flutter information**

<details>
<summary>flutter doctor</summary>

Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, 1.25.0-8.1.pre, on Microsoft Windows [Version 10.0.18363.1316],
    locale de-DE)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[√] Chrome - develop for the web
[√] Android Studio (version 4.1.0)
[√] VS Code (version 1.53.2)
[√] Connected device (1 available)

• No issues found!
</details>

### PR DESCRIPTION
**Pull request by vaind**
* and explicitly skips private fields.

closes #193

### CODE REVIEW

File: generator/integration-tests/final/1.dart [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        () => box.put(
+            A(0, 0, 0, '', 0, 0, [], Uint8List(0), [], false, Int8List(0))),
+        throwsA(predicate((ArgumentError e) =>
+            e.toString().contains('You must assign an ID'))));
+
+    final inserted = A(42, 1, 2, 'foo', 3, 4, [5, 6], Uint8List(1),
+        ['foo', 'bar'], true, Int8List(2));
+    box.put(inserted);
+    final read = box.get(inserted.id)!;
+    expect(read.tString, inserted.tString);
```
Comment (greenrobot-team) [inline generator/integration-tests/final/1.dart:60]: Why not verify all the other properties?

File: generator/lib/src/entity_resolver.dart [hunk 2]
```diff
@@ -99,6 +105,11 @@ class EntityResolver extends Builder {
         continue;
       }
 
+      if (f.name.startsWith('_')) {
```
Comment (greenrobot-team) [inline generator/lib/src/entity_resolver.dart:108]: Why not use `f.isPrivate`, is that different?

File: objectbox/lib/src/modelinfo/modelproperty.dart [hunk 3]
```diff
@@ -50,8 +50,25 @@ class ModelProperty {
 
   set dartFieldType(String value) => _dartFieldType = value;
 
-  String get fieldType =>
-      _dartFieldType!.replaceFirst('?', '', _dartFieldType!.length - 1);
+  // Used to inform about an ID field that is read-only (needed for code-gen).
+  // We're adding this prefix to [_dartFieldType] so that we don't have to do
```
Comment (greenrobot-team) [inline objectbox/lib/src/modelinfo/modelproperty.dart:54]: Update `_dartFieldType` docs to indicate this (also doesn't mention the `?`, yet)?

**Review by greenrobot-team — COMMENTED** [review]
Found some minor things, otherwise LGTM.

File: generator/integration-tests/final/1.dart (hunk 1 above)
Comment (vaind) [reply generator/integration-tests/final/1.dart:60]: I've added the rest

File: generator/lib/src/entity_resolver.dart (hunk 2 above)
Comment (vaind) [reply generator/lib/src/entity_resolver.dart:108]: missed that one. the implementation is the same so I've switched.

File: objectbox/lib/src/modelinfo/modelproperty.dart (hunk 3 above)
Comment (vaind) [reply objectbox/lib/src/modelinfo/modelproperty.dart:54]: updated the docs.

**Pull request merged**